File: Problems/0647. Palindromic Substrings/solution3.java
# Optimal Space + Time 
class Solution:
    def countSubstrings(self, s: str) -> int:
        result = 0

        for char in range(len(s)):
            l, r = char, char
            while l >= 0 and r < len(s) and s[l] == s[r]:
                result += 1
                l, r = l - 1, r + 1

            l, r = char, char + 1
            while l >= 0 and r < len(s) and s[l] == s[r]:
                result += 1
                l, r = l - 1, r + 1

        return result

File: Problems/1143. Longest Common Subsequence/solution3.py
# DP Optimal Time + Space
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        dp = [0] * (len(text2) + 1)
        
        for row in range(len(text1) - 1, -1, -1):
            prevDiag = 0
            for col in range(len(text2) - 1, -1, -1):
                temp = dp[col]
                if text1[row] == text2[col]:
                    dp[col] = 1 + prevDiag
                else:
                    dp[col] = max(dp[col], dp[col + 1]) 
                prevDiag = temp
        return dp[0]
